docs: ALTER TABLE COMPACT (#8722)
pingcap/docs: sql-statements/sql-statement-alter-table-compact.md

---
title: ALTER TABLE ... COMPACT
summary: An overview of the usage of ALTER TABLE ... COMPACT for the TiDB database.
---

# ALTER TABLE ... COMPACT

> **Warning:**
>
> This statement is still an experimental feature. It is NOT recommended that you use it in the production environment.

> After a write occurs, TiDB automatically performs data compaction at the backend. Specifically, TiDB rewrites the physical data in a table, including cleaning up deleted data and merging multiple versions of data. After data compaction, you can get higher access performance with less disk usage. The `ALTER TABLE ... COMPACT` statement allows you to compact data of specific tables immediately without waiting for the backend to trigger a compaction.

Reviewer: ```suggestion
After a write occurs, TiDB automatically performs data compaction in background. Specifically, TiDB rewrites the physical data, including cleaning up deleted rows and merging multiple versions caused by updates. When data compaction is finished, access performance can be improved and disk usage can be reduced. The `ALTER TABLE ... COMPACT` statement allows you to initiate a compaction for a specific table immediately, without waiting util triggered in background.
```

> TiDB automatically schedules data compaction on storage nodes in the background. During the compaction, storage nodes rewrite the physical data, including cleaning up deleted rows and merging multiple versions caused by updates. This enhances read performance with less disk usage. The `ALTER TABLE ... COMPACT` statement allows you to initiate compaction for a specific table immediately, without waiting until compaction is triggered in the background.

Reviewer: This is a nice and reasonable change! BTW how about using:

```
To enhance read performance and reduce disk usage, ...
```

as they do not have causal relationship.

> TiDB automatically schedules data compaction on storage nodes in the background. During the compaction, storage nodes rewrite the physical data, including cleaning up deleted rows and merging multiple versions caused by updates. This enhances read performance with less disk usage. The `ALTER TABLE ... COMPACT` statement allows you to initiate compaction for a specific table immediately, without waiting until compaction is triggered in the background.

Reviewer 2: LGTM

> The execution of this statement does not block the existing SQL statements, affect such TiDB functions as transactions, DDL, and GC, or change data obtained by executing SQL statements. However, executing this statement may take up some IO and CPU resources, which might result in business latency.

Reviewer: ```suggestion
The execution of this statement does not block existing SQL statements or affect any TiDB features, such as transactions, DDL, and GC. Data that can be selected via SQL statements will not be changed either. However, executing this statement may take up some IO and CPU resources, which might result in higher SQL execution latency.
```

> The execution finishes with execution result returned only when all replicas of a table are compacted. During the execution process, you can safely interrupt the compaction for the current table by executing the [`KILL`](/sql-statements/sql-statement-kill.md) statement. Interrupting a compaction does not break data consistency or lead to data loss, nor does it affect subsequent manual or automatic compactions.

Reviewer: ```suggestion
The compaction statement will finish the execution and return when all replicas of a table are compacted. During the execution process, you can safely interrupt the compaction by executing the [`KILL`](/sql-statements/sql-statement-kill.md) statement. Interrupting a compaction does not break data consistency or lead to data loss, nor does it affect subsequent manual or background compactions.
```

> Data compaction is currently supported only for TiFlash, not for TiKV.

Reviewer: ```suggestion
Data compaction is currently supported only for TiFlash replicas, not for TiKV replicas.
```

## Synopsis

```ebnf+diagram
AlterTableCompactStmt ::=
    'ALTER' 'TABLE' TableName 'COMPACT' 'TIFLASH' 'REPLICA'
```

## Examples

### Compact TiFlash replicas in a table

> Assume that an `employees` table has 4 partitions and 2 TiFlash replicas.

Reviewer: ```suggestion
Consider that an `employees` table has 4 partitions with 2 TiFlash replicas:
```

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
)
PARTITION BY LIST (store_id) (
    PARTITION pNorth VALUES IN (1, 2, 3, 4, 5),
    PARTITION pEast VALUES IN (6, 7, 8, 9, 10),
    PARTITION pWest VALUES IN (11, 12, 13, 14, 15),
    PARTITION pCentral VALUES IN (16, 17, 18, 19, 20)
);
ALTER TABLE employees SET TIFLASH REPLICA 2;
```

> You can execute the following statement to immediately compact data of 2 TiFlash replicas for all partitions in the `employees` table.

Reviewer: ```suggestion
You can execute the following statement to immediately initiate the compaction for the 2 TiFlash replicas for all partitions in the `employees` table:
```

{{< copyable "sql" >}}

```sql

> ALTER TABLE employee COMPACT TIFLASH REPLICA;

Reviewer: ```suggestion
ALTER TABLE employees COMPACT TIFLASH REPLICA;
```

```

## Concurrency

> The `ALTER TABLE ... COMPACT` statement compacts data of all replicas in a table at the same time.

Reviewer: ```suggestion
The `ALTER TABLE ... COMPACT` statement compacts all replicas in a table simultaneously.
```

> To avoid significant impact on online business, each TiFlash instance only processes compaction for one table at a time by default (automatic compaction in the backend is not affected). This means that if you execute the `ALTER TABLE ... COMPACT` statement on multiple tables at the same time, they will be queued for execution on the same TiFlash instance, rather than being executed simultaneously.

Reviewer: ```suggestion
To avoid significant impact on online business, each TiFlash instance only processes compaction command for one table at a time by default (except for the compaction triggered in the background). This means that if you execute the `ALTER TABLE ... COMPACT` statement on multiple tables at the same time, their execution will be queued on the same TiFlash instance, rather than being executed simultaneously.
```

> You can change the TiFlash configuration file parameter [`manual_compact_pool_size`](/tiflash/tiflash-configuration.md) to obtain greater table-level concurrency with higher resource usage. For example, if `manual_compact_pool_size` is set to 2, you can compact data of 2 tables simultaneously.

Reviewer: ```suggestion
To obtain greater table-level concurrency with higher resource usage, you can modify the TiFlash configuration [`manual_compact_pool_size`](/tiflash/tiflash-configuration.md). For example, when `manual_compact_pool_size` is set to 2, compaction for 2 tables will be processed simultaneously.
```

## MySQL compatibility

> The `ALTER TABLE ... COMPACT` syntax is an extension to the standard SQL syntax introduced by TiDB. Although there is no equivalent MySQL syntax, you can still execute this statement on MySQL clients of versions ranging from 5.7 to 8.0, or on various database drivers that follow the MySQL protocol.

Reviewer: ```suggestion
The `ALTER TABLE ... COMPACT` syntax is a TiDB specific syntax, which is an extension to the standard SQL syntax. Although there is no equivalent MySQL syntax, you can still execute this statement by using MySQL clients or various database drivers that follow the MySQL protocol.
```

## TiDB Binlog and TiCDC compatibility

> The `ALTER TABLE ... COMPACT` statement does not result in logical data changes and are therefore not replicated downstream by TiDB Binlog or TiCDC.

Reviewer: ```suggestion
The `ALTER TABLE ... COMPACT` statement does not result in logical data changes and are therefore not replicated to the downstreams by TiDB Binlog or TiCDC.
```
